doc/src/modules/solvers/solveset.rst

```diff
@@ -1,12 +1,463 @@
+  solution is returned. For input it only takes the equation, the variables
+  to solve for & the optional argument `domain` in which the equations has to
+  be solved.
+
+* ``solveset`` can return infinitely many solutions. For example solving for
+  `sin(x) = 0` returns `\{2 n \pi \| n \in \mathbb{Z}\} \cup \{2 n \pi + \pi \| n \in \mathbb{Z}\}`, whereas `solve`
+  only returns [0, π].
+
+* There is a clear code level and interface level separation between solvers
+  for equations in complex domain and equations in real domain. For example
+  solving `exp(x) == 1` when `x` is to be solved in complex `domain`, returns
+  the set of all solutions that is `\{2 n i \pi \| n \in \mathbb{Z}\}`, whereas if `x` is to be
+  solved in the Real `domain` then only `\{0\}` is returned.
+
+
+Why do we use Sets as an output type?
+-------------------------------------
+
+SymPy has a well developed sets module, which can represent most of the set
+containers in Mathematics such as:
+
+ * ``FiniteSet``
+   Represents a finite set of discrete numbers.
+
+ * ``Interval`` 
+   Represents a real interval as a Set.
+
+ * ``ProductSet``
+   Represents a Cartesian Product of Sets.
+
+ * ``ImageSet``
+   Represents the Image of a set under a mathematical function
+
+    >>> from sympy import ImageSet, S, Lambda
+    >>> from sympy.abc import x
+    >>> from sympy.sets import ImageSet
+    >>> squares = ImageSet(Lambda(x, x**2), S.Naturals)  # {x**2 for x in N}
+    >>> 4 in squares
+    True
+
+ * ``ComplexPlane``
+   Represents the Set of all Complex Numbers
+
+ * ``ConditionSet``
+   Represents the Set of elements which satisfies a given condition
+
+Also, the predefined set classes such as:
+
+ * ``Naturals`` `\mathbb{N}`
+   Represents the natural numbers (or counting numbers) which are all
+   positive integers starting from 1 
+
+ * ``Naturals0`` `\mathbb{W}`
+   Represents the whole numbers which are all the non-negative integers,
+   inclusive of zero
+
+ * ``Integers`` `\mathbb{Z}`
+   Represents all integers: positive, negative and zero. This set is also
+   available as the Singleton, ``S.Integers``.
+
+ * ``Reals`` `\mathbb{R}`
+   Represents the Set of all Real numbers.
+
+ * ``Complexes`` `\mathbb{C}`
+   Represents the Set of all Complex numbers.
+
+It is capable of most of the set operations in Mathematics:
+
+ * ``Union``
+ * ``Intersection``
+ * ``Complement``
+ * ``SymmetricDifference``
+
+The main reason for using sets as output to solvers is that it can consistently
+represent many types of solutions. For single variable case it can represent:
+
+ * No solution (by null set)
+
+ * Finitely many solutions (by FiniteSet)
+
+ * Infinitely many solutions, both countably and uncountably infinite solutions
+   (using the ImageSet module) Interval
+
+ * There can also be bizarre solutions to equations like set of all rational number.
+
+No other programmer's object (list, dictionary, python sets) provides the
```

Comment:
You should also add "generator" here, and maybe some words below. Someone asked about this at the SymPy BoF at SciPy. Not everyone gets that SymPy aims for all objects to be as symbolic as possible.